hotsos/core/plugintools.py

```diff
@@ -38,6 +38,197 @@ def yaml_dump(data):
                      default_flow_style=False).rstrip("\n")
 
 
+def markdown_expand_dict_key(data, level):
+    """Expand the data object.
+
+    Parameters:
+    data -- The data object. This can be a dict, list, or a flat type such as
+    int or str.
+    level -- The current header level
+
+    Returns:
+    A string expansion formatted in markdown of the data object.
+    """
+    level_prefix = ''.join(['#' for i in range(level)])
+    markdown_output = ''
+    if isinstance(data, (int, float, str)):
+        markdown_output += f'\n{data}\n'
+    elif isinstance(data, dict):
+        for key in data:
+            markdown_output += f'\n{level_prefix} {key}\n'
+            markdown_output += markdown_expand_dict_key(data[key], level + 1)
+    elif isinstance(data, list):
+        markdown_output += '\n'
+        for item in data:
+            markdown_output += f'- {item}\n'
+    return markdown_output
+
+
+def html_expand_dict_key(data, level, indent):
+    """Expand the data object.
+
+    Parameters:
+    data -- The data object. This can be a dict, list, or a flat type such as
+    int or str.
+    level -- The current header level
+    indent -- A string with spaces for indentation
+
+    Returns:
+    A string expansion formatted in markdown of the data object.
+    """
+    markup_output = ''
+    if isinstance(data, (int, float, str)):
+        markup_output += f'{indent}{data}\n'
+    elif isinstance(data, dict):
+        add_class = ''
+        if level == 1:
+            add_class = ' class="tree"'
+        markup_output += f'{indent}<ul{add_class}>\n'
+        for key in data.keys():
+            markup_output += f'{indent}    <li>\n' \
+                + f'{indent}        <details>\n' \
+                + f'{indent}            <summary>{key}</summary>\n'
+            markup_output += html_expand_dict_key(data[key], level + 1,
+                                                  indent + '            ')
+            markup_output += f'{indent}        </details>\n' \
+                + f'{indent}    </li>\n'
+        markup_output += f'{indent}</ul>\n'
+    elif isinstance(data, list):
+        markup_output += f'{indent}<ul>\n'
+        for item in data:
+            markup_output += f'{indent}    <li>\n'
+            markup_output += html_expand_dict_key(item, level,
+                                                  indent + '        ')
+            markup_output += f'{indent}    </li>\n'
+        markup_output += f'{indent}</ul>\n'
+    return markup_output
+
+
+def markdown_dump(data):
+    """Convert the data (dict) into a markdown document.
+
+    Parameters:
+    data: dict --- The data
+
+    Returns:
+    str -- The markdown document
+    """
+    markdown = '# hotsos summary\n'
```

Comment:
If you're stripping '\n' below do you need this here?

Reply:
Hmm good point. I'll fix that in the next iteration.

Reply:
I think I do if I don't want a trailing `\n`. The function `markdown_expand_dict_key()` recurses the `dict` and will return a string that has one too many `\n`.

Reply:
i think the collapsible sections are great but i think it is overkill to use them for everything. it would be really nice e.g. to have them e.g. just for the top-level keys.

Reply:
I added a `max_level` parameter to control how far down the output is collapsible.